### PR TITLE
Added fuzzy c-means

### PR DESCRIPTION
**Pull request by annimesh2809**
An implementation of the standard fuzzy c-means algorithm as stated [here](https://github.com/JuliaImages/ImageSegmentation.jl/pull/11). I have used [this](http://yaikhom.com/res/fcm/fcm.pdf) paper as a reference.

### CODE REVIEW

**Comment by ararslan** [comment]
Yes, for now I'd say you should make them 0.5 compatible. We can later drop 0.5 support and have FemtoCleaner handle the 0.5 -> 0.6 syntax changes.

**Comment by kmsquire** [comment]
@annimesh2809, looks good.  Could you add some tests?

**Comment by kmsquire** [comment]
@annimesh2809, thanks for adding tests. 

The tests are failing on v0.5 because `.=` isn't defined there, and on master because of a static compilation problem with StaticArrays. V0.6 isn't being tested by Travis. 

@ararslan, you seem pretty busy, but do you think you might be able to update Travis testing here?  I might be able to get to it this weekend, but not sure. 

After that's fixed, my opinion is that this package (and pull request) could be upgraded to only target v0.6 and later. It doesn't have any very active maintainers, and that just makes things easier. 

@annimesh2809, if you wanted it merged sooner, you could change all uses of `.=` to simply be `=` for now. Performance will probably go down for v0.6 and v0.7, but testing would probably pass. Or you can wait until this package is bumped to v0.6. You're also very welcome to bump the package to support only v0.6 and later yourself (in a separate pull request).

**Comment by annimesh2809** [comment]
@kmsquire I have updated the tests and now they pass successfully in Julia v0.5. I think it would be better to merge this first (as some more [PR's](https://github.com/JuliaImages/ImageSegmentation.jl/pull/11) depend on this now.) and later I will bump this package to v0.6 in a separate PR.

Also could you kindly take a look at issue #91. If you want, I can add the fix (i.e. change `kmeans(X::Matrix, ...)` to `kmeans{T<:AbstractFloat}(X::Matrix{T}, ...)` ) in this PR itself.

**Comment by kmsquire** [comment]
@annimesh2809, thank you. I hadn't realized that the v0.5 error was in the tests themselves. I'll merge this when I get to a computer (unless someone beats me to it). 

Regarding issue #91, can you please open up another pull request with a fix?  We ideally like to keep unrelated changes in different commits (at least) and generally different pull requests.

File: src/fuzzycmeans.jl [hunk 1]
```diff
@@ -0,0 +1,112 @@
+# Fuzzy C means algorithm
+
+## Interface
+
+immutable FuzzyCMeansResult{T<:AbstractFloat} <: ClusteringResult
+    centers::Matrix{T}          # cluster centers (d x C)
+    weights::Matrix{T}          # assigned weights (n x C)
```
Comment (kmsquire) [inline src/fuzzycmeans.jl]: If you make `centers` take on the type of the input `data`, then it's possible that `weights` will have a different type (e.g., in k-means, the centers have a parameterized type `T`, but the weight vector is a `Float64`).  How about doing the same thing here?

File: src/fuzzycmeans.jl [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+            end
+            weights[i,j] = 1.0/den
+        end
+    end
+end
+
+function update_centers!(centers, data, weights, fuzziness)
+    nrows, ncols = size(weights)
+    for j in 1:ncols
+        num = zeros(Float64, size(data[:,1]))
```
Comment (kmsquire) [inline src/fuzzycmeans.jl]: `data[:,1]` makes a copy here.  Use `size(data, 1)`.

Also, if you update `centers` below, then this should use `zeros(T, size(data,1))` (after getting the eltype of centers, of course).

File: src/fuzzycmeans.jl [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        den = 0.0
```
Comment (kmsquire) [inline src/fuzzycmeans.jl]: `den = zero(T)`

(Not actually sure this is still necessary--can someone say for sure?

File: src/fuzzycmeans.jl [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    weights ./= sum(weights, 2)
+
+    centers = fill(0.0, nrows, C)
+    prev_centers = identity.(centers)
+
+    δ = Inf
+    iter = 0
+
+    if displevel >= 2
+        prog = ProgressThresh(tol, "Converging:")
```
Comment (kmsquire) [inline src/fuzzycmeans.jl]: How slow is the algorithm here?  While I imagine `ProgressMeter` produces nice output , in most core packages like this, there is often a goal not to pull in too many external dependencies.  Is this output essential, or can you produce something reasonable without `ProgressMeter.jl`?

**Comment by annimesh2809** [comment]
@kmsquire I have replaced `ProgressMeter` with simple `printf` statements similar to `kmeans`. I have also made some other necessary changes.

**Comment by kmsquire** [comment]
@annimesh2809, could you add some documentation for this under docs (in a new pull request)?

**Comment by annimesh2809** [comment]
Yes, sure! But it might take some time because my midterms are coming. WIll that be okay?

**Comment by ararslan** [comment]
Yep, that's fine. Good luck on your midterms!

**Comment by timholy** [comment]
Would it be possible to get a new release?

**Comment by ararslan** [comment]
[The deed is done](https://github.com/JuliaLang/METADATA.jl/pull/10940).

**Pull request merged**